### PR TITLE
Convert onchange methods to stored-editable computed fields.

### PR DESCRIPTION
**Pull request by remybar**
Convert onchange methods to stored-editable computed fields for Service applications.
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 1]
```diff
@@ -40,10 +40,33 @@ def _domain_task_id(self):
...
+        'project.task',
+        'Task',
+        compute='_compute_task_id',
+        store=True,
+        readonly=False,
+        index=True,
+        domain="[('company_id', '=', company_id),"
+                "('project_id.allow_timesheets',"" '=', True),"
+                "('project_id', '=?', project_id)]")
+=======
```
Comment (LucasLefevre) [inline addons/hr_timesheet/models/hr_timesheet.py]: coucou conflict marker :laughing:

File: addons/hr_timesheet/models/project.py [hunk 2]
```diff
@@ -29,10 +36,11 @@ class Project(models.Model):
...
 
-    @api.onchange('analytic_account_id')
-    def _onchange_analytic_account(self):
-        if not self.analytic_account_id and self._origin:
-            self.allow_timesheets = False
+    @api.depends('analytic_account_id')
+    def _compute_allow_timesheets(self):
+        for record in self:
+            if not record.analytic_account_id:
+                record.allow_timesheets = False
```
Comment (LucasLefevre) [inline addons/hr_timesheet/models/project.py]: Could be batched
```python
@api.depends('analytic_account_id')
def _compute_allow_timesheets(self):
    without_account = self.filtered(lamdba t: not t.analytic_account_id)
    without_account.allow_timesheets = False
```

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 3]
```diff
@@ -62,17 +85,19 @@ def onchange_project_id(self):
             # reset task when changing project
             self.task_id = False
 
-    @api.onchange('task_id')
-    def _onchange_task_id(self):
-        if not self.project_id:
-            self.project_id = self.task_id.project_id
+    @api.depends('task_id.project_id')
+    def _compute_project_id(self):
+        for record in self:
```
Comment (LucasLefevre) [inline addons/hr_timesheet/models/hr_timesheet.py]:
```python
for timesheet in self: ...
```
Better to know what we are dealing with ;)
Everywhere.

File: addons/hr_timesheet/models/project.py [hunk 4]
```diff
@@ -8,7 +8,14 @@
...
     _inherit = "project.project"
 
-    allow_timesheets = fields.Boolean("Timesheets", default=True, help="Enable timesheeting on the project.")
+    allow_timesheets = fields.Boolean(
+        "Timesheets",
+        compute='_compute_allow_timesheets',
+        store=True,
+        readonly=False,
+        default=True,
+        help="Enable timesheeting on the project.")
```
Comment (LucasLefevre) [inline addons/hr_timesheet/models/project.py]: I know there aren't any guidelines about line splitting. But It is a bit too much IMHO

File: addons/project_timesheet_holidays/models/hr_holidays.py [hunk 5]
```diff
@@ -16,29 +16,42 @@ def _default_task_id(self):
...
+
+    @api.depends('timesheet_task_id', 'timesheet_project_id')
+    def _compute_timesheet_generate(self):
+        for holiday_type in self:
+            holiday_type.timesheet_generate = holiday_type.timesheet_task_id and \
+                holiday_type.timesheet_project_id
+
+    @api.depends('timesheet_project_id')
+    def _compute_timesheet_task_id(self):
+        for record in self:
```
Comment (LucasLefevre) [inline addons/project_timesheet_holidays/models/hr_holidays.py]: `for leave_type in self`

File: addons/project_timesheet_holidays/models/hr_holidays.py [hunk 6]
```diff
@@ -43,15 +44,15 @@ def _compute_timesheet_generate(self):
 
     @api.depends('timesheet_project_id')
     def _compute_timesheet_task_id(self):
-        for record in self:
-            company = record.company_id if record.company_id else record.env.company
+        for holidayType in self:
```
Comment (LucasLefevre) [inline addons/project_timesheet_holidays/models/hr_holidays.py:47]: use `snake_case`

File: addons/project/models/project.py [hunk 7]
```diff
@@ -705,6 +705,11 @@ def _compute_subtask_count(self):
         for task in self:
             task.subtask_count = len(self._get_all_subtasks())
 
+    @api.onchange('company_id')
+    def _onchange_task_company(self):
+        if self.project_id.company_id != self.company_id:
+            self.project_id = False
```
Comment (LucasLefevre) [inline addons/project/models/project.py]: Coucou new onchange in a task aimed at removing onchanges :laughing:

File: addons/project_timesheet_holidays/models/res_config_settings.py [hunk 8]
```diff
@@ -1,11 +1,16 @@
...
 class ResConfigSettings(models.TransientModel):
     _inherit = 'res.config.settings'
 
     leave_timesheet_project_id = fields.Many2one(related='company_id.leave_timesheet_project_id', string="Internal Project", domain="[('company_id', '=', company_id)]", readonly=False)
     leave_timesheet_task_id = fields.Many2one(related='company_id.leave_timesheet_task_id', string="Time Off Task", domain="[('company_id', '=', company_id)]", readonly=False)
+
+    @api.onchange('leave_timesheet_project_id')
+    def _on_change_timesheet_project_id(self):
+        if self.leave_timesheet_project_id != self.leave_timesheet_task_id.project_id:
+            self.leave_timesheet_task_id = False
```
Comment (LucasLefevre) [inline addons/project_timesheet_holidays/models/res_config_settings.py:16]: oO one more ^^

File: addons/hr_timesheet/models/hr_timesheet.py (hunk 1 above)
Comment (remybar) [reply addons/hr_timesheet/models/hr_timesheet.py]: Yep, but this one is already fixed I think ;-)

File: addons/hr_timesheet/models/project.py (hunk 2 above)
Comment (remybar) [reply addons/hr_timesheet/models/project.py]: done

File: addons/hr_timesheet/models/hr_timesheet.py (hunk 3 above)
Comment (remybar) [reply addons/hr_timesheet/models/hr_timesheet.py]: This was already done. I think you have reviewed an old version of this file.

File: addons/project_timesheet_holidays/models/hr_holidays.py (hunk 5 above)
Comment (remybar) [reply addons/project_timesheet_holidays/models/hr_holidays.py]: already done in the current version.

File: addons/project_timesheet_holidays/models/hr_holidays.py (hunk 6 above)
Comment (remybar) [reply addons/project_timesheet_holidays/models/hr_holidays.py:47]: done

File: addons/project/models/project.py (hunk 7 above)
Comment (remybar) [reply addons/project/models/project.py]: As explained in the review of the task #2169100, replacing this onchange by a compute field cause an infinite recursive loop on triggers. I think this onchange is part of a small list of onchange that need more thinking before removing them.

File: addons/project_timesheet_holidays/models/res_config_settings.py (hunk 8 above)
Comment (remybar) [reply addons/project_timesheet_holidays/models/res_config_settings.py:16]: The `leave_timesheet_task_id` field is a related field so I cannot convert it to a compute field. Then, this model is a transient one, that is just used to set some settings.
In this case, I think that it is acceptable to use an onchange as it is dedicated to a UI behavior.

File: addons/hr_timesheet/models/project.py (hunk 4 above)
Comment (remybar) [reply addons/hr_timesheet/models/project.py]: My point of view: putting 6 parameters on a same line or on 2 lines is not readable. When I cannot put all the parameters on a same line without ensuring a good readability, I prefer writing one parameter by line with the parameter name, so it's really easy to see without thinking which parameters are set and which are the values.

For me, almost all the field definitions in Odoo models are unreadable ... And I always keep in my that a code is written once but read a lot of times :-) But I'm open to a solution that takes care of readability more that compacity :-)

File: addons/hr_timesheet/models/hr_timesheet.py (hunk 1 above)
Comment (LucasLefevre) [reply addons/hr_timesheet/models/hr_timesheet.py]: But in two differents commits. Commit https://github.com/odoo/odoo/pull/45818/commits/f8d291dcca93820c8fb6ac7cc7c4b15be5a71ce4 introduces it and 2f1d69d removes it. We avoid introducing such noise in the commit history. Especially for conflict marker that should never appear in the history ;)

File: addons/project_timesheet_holidays/models/hr_holidays.py (hunk 5 above)
Comment (LucasLefevre) [reply addons/project_timesheet_holidays/models/hr_holidays.py]: same here IIRC, changed in two different commits

File: addons/project/models/project.py (hunk 7 above)
Comment (LucasLefevre) [reply addons/project/models/project.py]: Alright :)

File: addons/project/tests/test_project_flow.py [hunk 9]
```diff
@@ -96,32 +96,115 @@ def test_task_process_with_stages(self):
         self.assertEqual(task.stage_id.sequence, 1, "project_task: should have a stage with sequence=1")
 
     def test_subtask_process(self):
-        """ Check subtask mecanism and change it from project. """
+        """
+        Check subtask mecanism and change it from project.
```
Comment (tivisse) [inline addons/project/tests/test_project_flow.py:100]: Nice ! :+1:

**Comment by tivisse** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**